### ISSUE TITLE
** (Mix.Error) Can't continue due to errors on dependencies

### ISSUE THREAD

**Issue by andrelip** [JakeBecker/elixir-ls#71]
ElixirLS is giving me an exception on VS Code Problems window that is not happing anywhere else (not with mix compile or releasing using distillery). The system also runs fine.

 ** (Mix.Error) Can't continue due to errors on dependencies

Any tips on how to handle that problem? Is there a different elixir compile option so I can execute to try to reproduce that crash on terminal? 

### Environment

* Elixir & Erlang versions (elixir --version): 1.6
* Operating system: OSX 10.13.2
* Client (editor or IDE): Code 1.21.1

### Crash report template

an exception was raised:
    ** (Mix.Error) Can't continue due to errors on dependencies
        (mix) lib/mix.ex:291: Mix.raise/1
        (mix) lib/mix/tasks/deps.loadpaths.ex:27: Mix.Tasks.Deps.Loadpaths.run/1
        (mix) lib/mix/task.ex:314: Mix.Task.run_task/3
        (mix) lib/mix/tasks/loadpaths.ex:34: Mix.Tasks.Loadpaths.run/1
        (mix) lib/mix/task.ex:314: Mix.Task.run_task/3
        (mix) lib/mix/tasks/compile.ex:87: Mix.Tasks.Compile.run/1
        (mix) lib/mix/task.ex:314: Mix.Task.run_task/3
        (language_server) lib/language_server/build.ex:162: ElixirLS.LanguageServer.Build.compile/0

**Comment by norpan** [comment]
I had the exact same error. It went away when I removed the `deps` directory and let it fetch dependencies again.

**Comment by JakeBecker** [comment]
I'm guessing that running the task to `deps.get` before compiling is somehow screwing with the dependencies cache. I'm gonna have to play around with it a bit. Let me know if you figure out a reliable way to reproduce it.

**Comment by JakeBecker** [comment]
Until I fix this, try setting `elixirLS.fetchDeps` to `false` in your settings. If the problem was caused by the automatic fetching of dependencies, this might fix it.

**Comment by lukaszsamson** [comment]
Setting ```elixirLS.fetchDeps``` to ```false``` does not fix the issue. I run into it every time I upgrade erlang or elixir version.
Only ```rm -r deps _build .elixir_ls && mix deps.get``` helps.

**Comment by CharlotteDunois** [comment]
There is a similar error when there are dependencies with C source code and need a make compiler. On windows this can fail due to not being in the visual studio command prompt. Since this will not work, it should be possible to dismiss the problem/error.

**Comment by supernova32** [comment]
It also happens when adding a new dependency to `mix.exs`, or when updating them. It used to be fixable by simply quitting VSCode and opening again, but lately I need to remove `.elixir_ls/deps` in order to get the error to go away.

**Comment by AsharDweedar** [comment]
having the same issue here, for umbrella project.. but `rm -r deps _build .elixir_ls && mix deps.ge`  does not help in any way.

**Comment by bobics** [comment]
I actually have to quit VSCode before nuking the folders (per @lukaszsamson's comment), otherwise my deps only partially get fetched. It would leave my deps folder in borked state (hence the errors). Perhaps ElixirLS is locking a folder or there's a race condition. I think the magic sauce is:

1. Set `elixirLS.fetchDeps` to `false`.
2. Close VSCode just to be safe.
3. `rm -r deps _build .elixir_ls && mix deps.get`.
4. Start VSCode.

You should only have to do this once.

**Comment by michallepicki** [comment]
I can confirm that I had to close VSCode to get rid of this error (`rm -r deps _build .elixir_ls/build .elixir_ls/dialyzer_tmp` did not suffice). The error materialized itself in a situation where I already had `elixirLs.fetchDeps` set to `false`, and I switched from working on a branch with an added dependency to a branch without it.

**Comment by macbem** [comment]
This issue is in general related to the fact that `mix.exs` doesn't seem to be re-evaluated when an error happens in the file, even after changes. 

For example, I added a dependency as a list instead of a tuple, saved the file and got a `Dependency specified in the wrong format` error. I fixed the dependency, but the error is still there. It will fix itself once I trigger a VS Code window reload.

Maybe we should have a command to re-evaluate a file or restart the language service? For example, TypeScript LS allows us to restart it via a VS Code command and it fixes most issues with persisting errors.

**Comment by pietrofxq** [comment]
I just had this issue when specifying a version of a dep that was not compatible with my current version of elixir. Even after changing the version and refetching the deps and restarting vscode, the error wasn't gone. Only solution was to delete the `deps` folder and fetch again as mentioned above.

**Comment by jedeleh** [comment]
I had this issue and I found that simply restarting Atom fixed it. I didn't have to sacrifice a chicken or anything else more invasive.

**Comment by jedeleh** [comment]
I was able to make this stop happening by changing 2 configuration settings in ElixirLS:
1. uncheck Fetch Deps automatically option
2. changed Project Dir to .  (it was empty and had a red error alert around it)
When I restarted VScode the issue was resolved.

This does mean that you have to run mix deps.get and deps.compile manually when you make changes, but I kind of prefer that--I know the state of things in my conceptual understanding of my deps :)

**Comment by JakeBecker** [comment]
This project has moved!

It's now being maintained by proactive volunteers from the Elixir community over at [elixir-lsp](https://github.com/elixir-lsp)/[elixir-ls](https://github.com/elixir-lsp/elixir-ls). Updates will continue to be published from that repo to the original VS Code extension, so no need to switch plugins if you're using VS Code.

To avoid inundating the new maintainers with issues, please verify that your issue persists with the latest version of the extension (which is published from the new repo) before re-filing your issue there.

Thanks for using ElixirLS!

**Comment by jacmacmod** [comment]
> m -r deps _build .elixir_ls && mix deps.get

This did not help

**Comment by Deankinyua** [comment]
I was having this issue ..............deleting the whole .elixir_ls directory and the mix.lock file worked for me.

**Comment by miguno** [comment]
I had a very similar problem and found my way here via a Google search.

## My solution

For others reading this, here's what I needed to do:

1. Stop your IDE/editor including the elixir LSP.
2. Kill any left-over elixir/elixir-ls processes. (This was important.) Find them via e.g. `procs elixir` or `ps axu | grep elixir`.
3. In your project directory: `rm -rf .elixir_ls _build mix.lock && mix deps.get`.
4. Restart your IDE/editor including the elixir LSP.

## Environment

- elixir-ls v0.23.0
- Elixir 1.17.2 (compiled with Erlang/OTP 27)
- Erlang/OTP 27 [erts-15.0] [source] [64-bit] [smp:10:10] [ds:10:10:10] [async-threads:1] 

## Error message

The error message I originally encountered was (first few lines):

```
[ERROR][2024-09-10 20:08:13] ...lsp/handlers.lua:623    "Process #PID<0.287.0> raised an exception\n** (Mix.Error) Can't continue due to errors on dependencies\n    (mix 1.17.2) lib/mix.ex:588: Mix.raise/2\n    (m
ix 1.17.2) lib/mix/tasks/deps.loadpaths.ex:48: Mix.Tasks.Deps.Loadpaths.run/1\n    (mix 1.17.2) lib/mix/task.ex:495: anonymous fn/3 in Mix.Task.run_task/5\n    (mix 1.17.2) lib/mix/tasks/loadpaths.ex:37: Mix.Tasks
.Loadpaths.run/1\n    (mix 1.17.2) lib/mix/task.ex:495: anonymous fn/3 in Mix.Task.run_task/5\n    (mix 1.17.2) lib/mix/tasks/compile.ex:153: Mix.Tasks.Compile.run/1\n    (mix 1.17.2) lib/mix/task.ex:495: anonymou
s fn/3 in Mix.Task.run_task/5\n    (language_server 0.23.0) lib/language_server/build.ex:442: ElixirLS.LanguageServer.Build.run_mix_compile/1\n    (language_server 0.23.0) lib/language_server/build.ex:80: anonymou
s fn/3 in ElixirLS.LanguageServer.Build.build/3\n    (stdlib 6.0) timer.erl:590: :timer.tc/2\n    (language_server 0.23.0) lib/language_server/build.ex:20: anonymous fn/3 in ElixirLS.LanguageServer.Build.build/3\n
    (kernel 10.0) global.erl:699: :global.trans/4"
```

**Comment by jdsampayo** [comment]
@miguno's solution worked for me on Cursor. For step 2, I recommend the following:

1. Check which processes are running: `pkill -fl elixir` (Optional)
2. Terminate them: `pkill -9 -f elixir`
3. Verify that they were terminated: `pkill -fl elixir` (Optional)

I noticed that if I do not pass -9, it uses -15 as the default signal. However, upon rechecking, I found that the processes were never terminated at all with the default option.
